kratos/future/input_output/vtu_output.cpp

```diff
@@ -0,0 +1,1689 @@
+    switch (rLocation) {
+        case Globals::DataLocation::NodeHistorical:
+        case Globals::DataLocation::NodeNonHistorical:
+            found_existing_name = rUnstructuredGridData.mMapOfPointTensorAdaptors.find(rName) != rUnstructuredGridData.mMapOfPointTensorAdaptors.end();
+            break;
+        case Globals::DataLocation::Condition:
+        case Globals::DataLocation::Element:
+            found_existing_name = rUnstructuredGridData.mMapOfCellTensorAdaptors.find(rName) != rUnstructuredGridData.mMapOfCellTensorAdaptors.end();
+            break;
+        default:
+            KRATOS_ERROR << "Unsupported data location type.";
+    }
+    KRATOS_ERROR_IF(found_existing_name)
+            << "Found an existing data array with the same name = \"" << rName << "\".\n";
+}
+
+void CheckDataArrayName(
+    const std::string& rName,
+    const Globals::DataLocation& rLocation,
+    const std::vector<VtuOutput::UnstructuredGridData>& rListOfUnstructuredGridData)
+{
+    for (const auto& r_model_part_data : rListOfUnstructuredGridData) {
+        switch (rLocation) {
+            case Globals::DataLocation::NodeHistorical:
+            case Globals::DataLocation::NodeNonHistorical:
+                if (r_model_part_data.UsePointsForDataFieldOutput) {
+                    CheckDataArrayName(rName, rLocation, r_model_part_data);
+                }
+                break;
+            default:
+                CheckDataArrayName(rName, rLocation, r_model_part_data);
+        }
+    }
+}
+
+std::string GetEntityName(const std::optional<VtuOutput::CellContainerPointerType>& pCellContainer)
+{
+    if (pCellContainer.has_value()) {
+        return std::visit([](auto p_cell_container) {
+            using container_type = BareType<decltype(*p_cell_container)>;
+            return ModelPart::Container<container_type>::GetEntityName();
+        }, pCellContainer.value());
+    } else {
+        return "node";
+    }
+}
+
+void CopyAttributes(
+    const XmlElement& rSource,
+    XmlElement& rDestination)
+{
+    for (const auto& [attribute, value] : rSource.GetAttributes()) {
+        rDestination.AddAttribute(attribute, value);
+    }
+}
+
+template<class NDDataPointerType>
+NDDataPointerType GetWritingNDData(
+    const std::vector<IndexType>& rWritingIndices,
+    NDDataPointerType pNDData)
+{
+    KRATOS_TRY
+
+    const auto& origin_shape = pNDData->Shape();
+
+    KRATOS_ERROR_IF(origin_shape.size() == 0)
+        << "NDData shape should have at least one dimension representing number of entities [ nd data = "
+        << *pNDData << " ].\n";
+
+    if (rWritingIndices.size() == origin_shape[0]) {
+        // number of writing indices are same as the first dimension of the NdData, hence
+        // there is nothing to be ignored from this data set. Returning the original
+        return pNDData;
+    } else {
+        // now there is a mismatch between number of writing indices and the number of entities represented by the pNDData.
```

Comment:
this else is for the case at which you retain some indices. This corresponds to something on the line of 

~~~py
mask = retained_indeces
return pNDData[mask]
~~~

wouldn't it make sense to have a function "ReturnEntriesByMask" to do exactly this? if you accept an mdspan as input it should not be difficult to write ...